datalad/customremotes/main.py

```diff
@@ -75,5 +74,12 @@ def main(args=None, cls=None, remote_name=None, description=None):
         try:
             _main(args, cls)
         except Exception as exc:
-            lgr.error('%s (%s)', str(exc), exc.__class__.__name__)
+            # `SpecialRemote` classes are supposed to catch everything and
+            # turn it into a `RemoteError` resulting in an ERROR message to
+            # annex. If we end up here, something went wrong outside of the
+            # `master.Listen()` call in `_main`.
+            # In any case, exiting the special remote process should be
+            # accompanied by such an ERROR message to annex rather than a log
+            # message.
+            print("ERROR %s (%s)" % (str(exc), exc.__class__.__name__))
```

Comment:
I would say we better have both since if git-annex just swallows that ERROR and might die some other way (may be it was the first one to be killed?) then we might not see this error anywhere... Greedy me would keep the log (it will be the main datalad logger right?) and  channel ERROR to annex.   or am I missing the whole point? ;)

Reply:
Well, several things, @yarikoptic:

1. I have never seen annex swallow those. And if it does, that would be something to address in annex from my POV. Special remote protocol explicitly expects this kind of message and considers it a violation to exit non-zero w/o it. If anything, datalad swallows it by not considering stderr of the annex process it calls.
2. Within a datalad context where we don't swallow errors from annex (which should be most cases), this would lead to double reporting again, because what do we do with an error message we got from annex? We log it.

I think having both only makes sense when we finally manage to log to files by default and separate logging from user output.

Reply:
implemented the promise in 1441891f7f4c163b2fcad234a3288c4ddc738e66 , accompanied with a test for that block (was not unittested at all)